File: 13_Stack & Queue/E-TUF-Implement Stack using Arrays.java
// Implement stack using array (LIFO)
// https://takeuforward.org/plus/dsa/problems/implement-stack-using-arrays

class ArrayStack {
    private int[] arr;    // array to store elements
    private int top;      // index of the top element
    private int capacity; // maximum capacity of the stack

    // Constructor
    public ArrayStack() {
        capacity = 1000; // default capacity (can be adjusted)
        arr = new int[capacity];
        top = -1; // stack starts empty
    }

    // Push element onto stack
    public void push(int x) {
        if (top == capacity - 1) {
            throw new RuntimeException("Stack overflow");
        }
        arr[++top] = x;
    }

    // Pop element from stack
    public int pop() {
        if (isEmpty()) {
            throw new RuntimeException("Stack underflow");
        }
        return arr[top--];
    }

    // Get top element without removing it
    public int top() {
        if (isEmpty()) {
            throw new RuntimeException("Stack is empty");
        }
        return arr[top];
    }

    // Check if stack is empty
    public boolean isEmpty() {
        return top == -1;
    }

    // Driver test
    public static void main(String[] args) {
        ArrayStack stack = new ArrayStack();
        stack.push(5);
        stack.push(10);
        System.out.println(stack.top());     // Output: 10
        System.out.println(stack.pop());     // Output: 10
        System.out.println(stack.isEmpty()); // Output: false
    }
}

File: JavaScript/13_Stack & Queue/E-TUF-Implement Stack using Arrays.js
// Implement Stack using Arrays (LIFO)
// https://takeuforward.org/plus/dsa/problems/implement-stack-using-arrays

/**
 * Creates a stack using array (functional approach)
 * @param {number} capacity - Maximum capacity of the stack
 * @returns {Object} Stack operations
 */
function createArrayStack(capacity = 1000) {
    const arr = new Array(capacity);
    let top = -1;

    function push(x) {
        if (top === capacity - 1) {
            throw new Error("Stack overflow");
        }
        arr[++top] = x;
    }

    function pop() {
        if (isEmpty()) {
            throw new Error("Stack underflow");
        }
        return arr[top--];
    }

    function peek() {
        if (isEmpty()) {
            throw new Error("Stack is empty");
        }
        return arr[top];
    }

    function isEmpty() {
        return top === -1;
    }

    function size() {
        return top + 1;
    }

    return { push, pop, peek, isEmpty, size };
}

// Test cases
const stack = createArrayStack();
stack.push(5);
stack.push(10);
console.log(stack.peek()); // 10
console.log(stack.pop()); // 10
console.log(stack.isEmpty()); // false
console.log(stack.size()); // 1
stack.push(20);
stack.push(30);
console.log(stack.pop()); // 30
console.log(stack.peek()); // 20
